File: src/pipeline/context.rs
use std::collections::HashSet;
use std::path::{Path, PathBuf};

use indexmap::IndexMap;

use crate::config::{Config, ServiceOverride};

/// All data needed to run the create workspace pipeline.
/// Decoupled from App — can be built from Config alone (for CLI).
pub struct CreateContext {
    pub workspace_name: String,
    pub branch: String,
    pub config: Config,
    pub config_dir: PathBuf,
    pub session: String,
    pub unique_dirs: Vec<String>,
    pub dir_paths: Vec<(String, String)>,
    pub dir_branches: Vec<(String, String)>,
    pub shared_overrides: Vec<(String, IndexMap<String, ServiceOverride>, Vec<String>)>,
    pub bind_ip: String,
    pub skip_stages: HashSet<usize>,
    /// Per-repo target branches (dir_name → branch). If set, overrides ctx.branch for each dir.
    pub selected_dirs: Option<Vec<(String, String)>>,
}

impl CreateContext {
    /// Build context from Config (for CLI — no App needed).
    pub fn from_config(
        config: &Config,
        config_path: &Path,
        ws_name: &str,
        branch: &str,
        skip_stages: HashSet<usize>,
    ) -> anyhow::Result<Self> {
        let workspaces = config.all_workspaces();
        let entries = workspaces.get(ws_name)
            .ok_or_else(|| anyhow::anyhow!("workspace '{}' not found", ws_name))?;

        let mut unique_dirs = Vec::new();
        for entry in entries {
            if let Ok((dir, _)) = config.find_service_entry(entry) {
                if !unique_dirs.contains(&dir) {
                    unique_dirs.push(dir);
                }
            }
        }
        if unique_dirs.is_empty() {
            anyhow::bail!("no dirs found in workspace '{}'", ws_name);
        }

        let config_dir = config_path.parent().unwrap_or(Path::new(".")).to_path_buf();
        let default_branch = config.global_default_branch();

        // Resolve dir paths (through main workspace folder)
        let dir_paths: Vec<(String, String)> = unique_dirs.iter()
            .map(|d| {
                let p = Path::new(d);
                let resolved = if p.is_absolute() {
                    d.to_string()
                } else {
                    // Try workspace folder first, fallback to config dir
                    let ws_path = config_dir.join(format!("workspace--{default_branch}")).join(d);
                    if ws_path.exists() {
                        ws_path.to_string_lossy().into_owned()
                    } else {
                        config_dir.join(d).to_string_lossy().into_owned()
                    }
                };
                (d.clone(), resolved)
            })
            .collect();

        // Resolve dir branches (current git branch for each dir)
        let dir_branches: Vec<(String, String)> = unique_dirs.iter()
            .map(|d| {
                let dir_path = dir_paths.iter()
                    .find(|(name, _)| name == d)
                    .map(|(_, p)| p.as_str())
                    .unwrap_or(".");
                let branch = git_branch(dir_path).unwrap_or_else(|| "main".to_string());
                (d.clone(), branch)
            })
            .collect();

        // Resolve shared overrides for each dir
        let shared_overrides: Vec<(String, IndexMap<String, ServiceOverride>, Vec<String>)> =
            unique_dirs.iter()
                .map(|d| {
                    let (ov, hosts) = resolve_shared_overrides(config, d);
                    (d.clone(), ov, hosts)
                })
                .collect();

        // IP will be allocated in Provision stage if not provided
        let bind_ip = String::new();

        Ok(Self {
            workspace_name: ws_name.to_string(),
            branch: branch.to_string(),
            config: config.clone(),
            config_dir,
            session: config.session.clone(),
            unique_dirs,
            dir_paths,
            dir_branches,
            shared_overrides,
            bind_ip,
            skip_stages,
            selected_dirs: None,
        })
    }

    /// Build context with specific repo selection (from TUI checklist).
    #[allow(dead_code)]
    pub fn from_config_with_selection(
        config: &Config,
        config_path: &Path,
        ws_name: &str,
        branch: &str,
        selected: Vec<(String, String)>, // (dir_name, target_branch)
    ) -> anyhow::Result<Self> {
        let mut ctx = Self::from_config(config, config_path, ws_name, branch, HashSet::new())?;
        // Filter unique_dirs to only selected ones
        let selected_dir_names: Vec<String> = selected.iter().map(|(d, _)| d.clone()).collect();
        ctx.unique_dirs.retain(|d| selected_dir_names.contains(d));
        ctx.dir_paths.retain(|(d, _)| selected_dir_names.contains(d));
        ctx.dir_branches.retain(|(d, _)| selected_dir_names.contains(d));
        ctx.shared_overrides.retain(|(d, _, _)| selected_dir_names.contains(d));
        ctx.selected_dirs = Some(selected);
        Ok(ctx)
    }
}

/// All data needed to run the delete workspace pipeline.
#[allow(dead_code)]
pub struct DeleteContext {
    pub branch: String,
    pub config: Config,
    pub config_dir: PathBuf,
    pub session: String,
    pub wt_keys: Vec<String>,
    pub cleanup_items: Vec<CleanupItem>,
    pub dbs_to_drop: Vec<DbDropItem>,
    pub network: String,
    pub skip_stages: HashSet<usize>,
}

pub struct CleanupItem {
    pub dir_path: String,
    pub wt_path: PathBuf,
    pub wt_branch: String,
    pub pre_delete: Vec<String>,
}

pub struct DbDropItem {
    pub host: String,
    pub port: u16,
    pub db_name: String,
    pub user: String,
    pub password: String,
}

// ── Standalone helpers (extracted from App) ──

/// Resolve shared service overrides for a dir.
/// Returns (service_overrides merged with shared disabled profiles, shared hostnames).
pub fn resolve_shared_overrides(
    config: &Config,
    dir_name: &str,
) -> (IndexMap<String, ServiceOverride>, Vec<String>) {
    let dir = match config.repos.get(dir_name) {
        Some(d) => d,
        None => return (Default::default(), Vec::new()),
    };
    let wt_cfg = match dir.wt() {
        Some(wt) => wt,
        None => return (Default::default(), Vec::new()),
    };
    let mut overrides = wt_cfg.service_overrides.clone();
    let mut hosts: Vec<String> = Vec::new();

    for sref in &wt_cfg.shared_services {
        if !overrides.contains_key(&sref.name) {
            overrides.insert(sref.name.clone(), ServiceOverride {
                environment: IndexMap::new(),
                profiles: vec!["disabled".to_string()],
                mem_limit: None,
            });
        }
        if let Some(svc_def) = config.shared_services.get(&sref.name) {
            if let Some(host) = &svc_def.host {
                if !hosts.contains(host) {
                    hosts.push(host.clone());
                }
            }
        }
    }
    // Add proxy hostnames for all repos with proxy_port (so Docker can reach them via host-gateway)
    for (_, repo) in &config.repos {
        if let Some(port) = repo.proxy_port {
            if let Some(alias) = &repo.alias {
                // Proxy hostname will be resolved by the proxy on 127.0.0.1
                // Docker containers reach it via host-gateway → 127.0.0.1:port → proxy → bind_ip:port
                let _ = port; // port used for route registration, not for extra_hosts
                let hostname = format!("{alias}.tncli.test");
                if !hosts.contains(&hostname) {
                    hosts.push(hostname);
                }
            }
        }
    }

    (overrides, hosts)
}

/// Get current git branch for a directory path.
fn git_branch(dir_path: &str) -> Option<String> {
    let output = std::process::Command::new("git")
        .args(["-C", dir_path, "rev-parse", "--abbrev-ref", "HEAD"])
        .output()
        .ok()?;
    if !output.status.success() { return None; }
    let branch = String::from_utf8_lossy(&output.stdout).trim().to_string();
    if branch.is_empty() { None } else { Some(branch) }
}

File: internal/pipeline/context.go
package pipeline

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"

	"github.com/toantran292/tncli/internal/config"
)

// CreateContext holds all data needed for workspace creation pipeline.
type CreateContext struct {
	WorkspaceName   string
	Branch          string
	Config          *config.Config
	ConfigDir       string
	Session         string
	TmuxSession     string
	UniqueDirs      []string
	DirPaths        [][2]string // (dir_name, resolved_path)
	DirBranches     [][2]string // (dir_name, current_branch)
	SharedOverrides []SharedOverrideEntry
	BindIP          string
	SkipStages      map[int]bool
	SelectedDirs    [][2]string // (dir_name, target_branch) — nil if not set
}

type SharedOverrideEntry struct {
	DirName   string
	Overrides map[string]*config.ServiceOverride
	Hosts     []string
}

// DeleteContext holds all data needed for workspace deletion pipeline.
type DeleteContext struct {
	Branch       string
	Config       *config.Config
	ConfigDir    string
	CleanupItems []CleanupItem
	DBsToDrop    []DBDropItem
	Network      string
	SkipStages   map[int]bool
}

type CleanupItem struct {
	DirPath   string
	WtPath    string
	WtBranch  string
	PreDelete []string
}

type DBDropItem struct {
	Host     string
	Port     uint16
	DBName   string
	User     string
	Password string
}

// FromConfig builds CreateContext from Config.
func FromConfig(cfg *config.Config, configPath, wsName, branch string, skipStages map[int]bool) (*CreateContext, error) {
	workspaces := cfg.AllWorkspaces()
	entries, ok := workspaces[wsName]
	if !ok {
		return nil, fmt.Errorf("workspace '%s' not found", wsName)
	}

	var uniqueDirs []string
	seen := make(map[string]bool)
	for _, entry := range entries {
		d, _, err := cfg.FindServiceEntry(entry)
		if err != nil {
			continue
		}
		if !seen[d] {
			seen[d] = true
			uniqueDirs = append(uniqueDirs, d)
		}
	}
	if len(uniqueDirs) == 0 {
		return nil, fmt.Errorf("no dirs found in workspace '%s'", wsName)
	}

	configDir := filepath.Dir(configPath)
	defaultBranch := cfg.GlobalDefaultBranch()

	// Resolve dir paths
	var dirPaths [][2]string
	for _, d := range uniqueDirs {
		resolved := d
		if !filepath.IsAbs(d) {
			wsPath := filepath.Join(configDir, "workspace--"+defaultBranch, d)
			if isDir(wsPath) {
				resolved = wsPath
			} else {
				resolved = filepath.Join(configDir, d)
			}
		}
		dirPaths = append(dirPaths, [2]string{d, resolved})
	}

	// Resolve dir branches
	var dirBranches [][2]string
	for _, d := range uniqueDirs {
		dirPath := ""
		for _, dp := range dirPaths {
			if dp[0] == d {
				dirPath = dp[1]
				break
			}
		}
		b := gitBranch(dirPath)
		if b == "" {
			b = "main"
		}
		dirBranches = append(dirBranches, [2]string{d, b})
	}

	// Resolve shared overrides
	var sharedOverrides []SharedOverrideEntry
	for _, d := range uniqueDirs {
		ov, hosts := ResolveSharedOverrides(cfg, d)
		sharedOverrides = append(sharedOverrides, SharedOverrideEntry{DirName: d, Overrides: ov, Hosts: hosts})
	}

	return &CreateContext{
		WorkspaceName:   wsName,
		Branch:          branch,
		Config:          cfg,
		ConfigDir:       configDir,
		Session:         cfg.Session,
		TmuxSession:     cfg.SvcSession(),
		UniqueDirs:      uniqueDirs,
		DirPaths:        dirPaths,
		DirBranches:     dirBranches,
		SharedOverrides: sharedOverrides,
		BindIP:          "",
		SkipStages:      skipStages,
	}, nil
}

// FromConfigWithSelection builds CreateContext with specific repo selection.
func FromConfigWithSelection(cfg *config.Config, configPath, wsName, branch string, selected [][2]string) (*CreateContext, error) {
	ctx, err := FromConfig(cfg, configPath, wsName, branch, nil)
	if err != nil {
		return nil, err
	}

	selectedNames := make(map[string]bool)
	for _, s := range selected {
		selectedNames[s[0]] = true
	}

	// Filter
	var filteredDirs []string
	for _, d := range ctx.UniqueDirs {
		if selectedNames[d] {
			filteredDirs = append(filteredDirs, d)
		}
	}
	ctx.UniqueDirs = filteredDirs

	var filteredPaths [][2]string
	for _, dp := range ctx.DirPaths {
		if selectedNames[dp[0]] {
			filteredPaths = append(filteredPaths, dp)
		}
	}
	ctx.DirPaths = filteredPaths

	var filteredBranches [][2]string
	for _, db := range ctx.DirBranches {
		if selectedNames[db[0]] {
			filteredBranches = append(filteredBranches, db)
		}
	}
	ctx.DirBranches = filteredBranches

	var filteredOverrides []SharedOverrideEntry
	for _, so := range ctx.SharedOverrides {
		if selectedNames[so.DirName] {
			filteredOverrides = append(filteredOverrides, so)
		}
	}
	ctx.SharedOverrides = filteredOverrides

	ctx.SelectedDirs = selected
	ctx.SkipStages = nil
	return ctx, nil
}

// ResolveSharedOverrides resolves shared service overrides for a dir.
func ResolveSharedOverrides(cfg *config.Config, dirName string) (map[string]*config.ServiceOverride, []string) {
	dir, ok := cfg.Repos[dirName]
	if !ok || dir.WT() == nil {
		return nil, nil
	}
	wt := dir.WT()
	overrides := make(map[string]*config.ServiceOverride)
	for k, v := range wt.ServiceOverrides {
		overrides[k] = v
	}
	var hosts []string

	for _, sref := range wt.SharedServices {
		if _, ok := overrides[sref.Name]; !ok {
			overrides[sref.Name] = &config.ServiceOverride{
				Profiles: []string{"disabled"},
			}
		}
		host := cfg.SharedHost(sref.Name)
		if !contains(hosts, host) {
			hosts = append(hosts, host)
		}
	}

	for _, svcName := range wt.Disable {
		if _, ok := overrides[svcName]; !ok {
			overrides[svcName] = &config.ServiceOverride{
				Profiles: []string{"disabled"},
			}
		}
	}

	for _, repo := range cfg.Repos {
		if repo.ProxyPort != nil && repo.Alias != "" {
			hostname := repo.Alias + ".tncli.test"
			if !contains(hosts, hostname) {
				hosts = append(hosts, hostname)
			}
		}
	}

	return overrides, hosts
}

func gitBranch(dirPath string) string {
	out, err := exec.Command("git", "-C", dirPath, "rev-parse", "--abbrev-ref", "HEAD").Output()
	if err != nil {
		return ""
	}
	return strings.TrimSpace(string(out))
}

func isDir(path string) bool {
	info, err := os.Stat(path)
	return err == nil && info.IsDir()
}

func contains(ss []string, s string) bool {
	for _, v := range ss {
		if v == s {
			return true
		}
	}
	return false
}
